File: internal/agentrun/activity_test.go
package agentrun

import "testing"

func TestDeriveActivityStateDecisionTable(t *testing.T) {
	tests := []struct {
		name         string
		mutateRun    func(*Run)
		intelligence IntelligenceState
		want         ActivityState
	}{
		{name: "active defaults to running", want: ActivityRunning},
		{name: "active working", intelligence: IntelligenceWorking, want: ActivityRunning},
		{name: "active drift signal remains running", intelligence: IntelligencePotentiallyDrifting, want: ActivityRunning},
		{name: "waiting comes from intelligence", intelligence: IntelligenceWaiting, want: ActivityWaiting},
		{name: "blocked comes from intelligence", intelligence: IntelligenceBlocked, want: ActivityBlocked},
		{name: "explicit completion", intelligence: IntelligenceCompleted, want: ActivityCompleted},
		{name: "explicit failure", intelligence: IntelligenceFailed, want: ActivityFailed},
		{
			name: "stale lifecycle takes precedence", mutateRun: func(run *Run) {
				run.State = StateStale
				run.LeaseState = LeaseExpired
			},
			intelligence: IntelligenceFailed, want: ActivityStale,
		},
		{
			name: "successful exit stays unknown", mutateRun: func(run *Run) {
				run.State = StateExited
				run.LeaseState = LeaseExpired
				run.Exit = &Exit{Code: 0, Result: "completed"}
			},
			want: ActivityUnknown,
		},
		{
			name: "nonzero exit is failed through intelligence", mutateRun: func(run *Run) {
				run.State = StateExited
				run.LeaseState = LeaseExpired
				run.Exit = &Exit{Code: 2, Result: "failed"}
			},
			intelligence: IntelligenceFailed, want: ActivityFailed,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			run := intelligenceRun()
			if test.mutateRun != nil {
				test.mutateRun(&run)
			}
			got := DeriveActivityState(run, RunIntelligence{State: test.intelligence})
			if got != test.want {
				t.Fatalf("activity = %q, want %q", got, test.want)
			}
		})
	}
}

File: crates/ptrack-agent/src/activity_test.rs
use super::{
    ActivityState, IntelligenceConfidence, IntelligenceState, LeaseState, ProcessState,
    RegistrationKind, Run, RunIntelligence, RunState, derive_activity_state,
};

fn intelligence(state: IntelligenceState) -> RunIntelligence {
    RunIntelligence {
        run_id: "run-1".to_owned(),
        state,
        confidence: IntelligenceConfidence::Medium,
        evidence: Vec::new(),
        event_count: 0,
        last_event_at: super::Timestamp::ZERO,
    }
}

#[test]
fn activity_mapping_is_conservative() {
    let active = Run {
        registration_kind: RegistrationKind::Launched,
        state: RunState::Running,
        process_state: ProcessState::Running,
        ..Run::default()
    };
    for (state, expected) in [
        (IntelligenceState::Unknown, ActivityState::Running),
        (IntelligenceState::Waiting, ActivityState::Waiting),
        (IntelligenceState::Blocked, ActivityState::Blocked),
        (IntelligenceState::Completed, ActivityState::Completed),
        (IntelligenceState::Failed, ActivityState::Failed),
    ] {
        assert_eq!(
            derive_activity_state(&active, &intelligence(state)),
            expected
        );
    }
    let inactive = Run {
        registration_kind: RegistrationKind::External,
        state: RunState::Exited,
        lease_state: LeaseState::Expired,
        ..Run::default()
    };
    assert_eq!(
        derive_activity_state(&inactive, &intelligence(IntelligenceState::Waiting)),
        ActivityState::Unknown
    );
    let stale = Run {
        state: RunState::Stale,
        ..active
    };
    assert_eq!(
        derive_activity_state(&stale, &intelligence(IntelligenceState::Failed)),
        ActivityState::Stale
    );
}
